lib/tree.rb

```diff
@@ -18,45 +18,101 @@ def initialize
 
   # Time Complexity: 
   # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  def add(key, value, current = @root)
+    new_node = TreeNode.new(key, value)
+
+    if @root == nil
+      @root = new_node
+    else
+      if key <= current.key
+
+        if current.left == nil
+          current.left = new_node
+        else
+          add(key, value, current.left)
+        end
+
+      else
+
+        if current.right == nil
+          current.right = new_node
+        else
+          add(key, value, current.right)
+        end
+
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def find(key)
-    raise NotImplementedError
+  # Time Complexity:
+  # Space Complexity:
+  def find(key, current = @root)
+    if @root == nil
+      return nil
+    elsif key == current.key
+      return current.value
+    elsif key <= current.key
+      find(key, current.left)
+    else
+      find(key, current.right)
+    end
   end
 
   # Time Complexity: 
   # Space Complexity: 
-  def inorder
-    raise NotImplementedError
+  def inorder(current = @root, array = [])
+    if current == nil
+      return array
+    else
+      inorder(current.left, array)
+      array << { key: current.key, value: current.value }
+      inorder(current.right, array)
+    end
   end
 
+
+
   # Time Complexity: 
   # Space Complexity: 
-  def preorder
-    raise NotImplementedError
+  def preorder(current = @root, array = [])
+    if current  == nil
+      return array
+    else
+      array << { key: current.key, value: current.value }
+      preorder(current.left, array)
+      preorder(current.right, array)
+    end
   end
 
   # Time Complexity: 
   # Space Complexity: 
-  def postorder
-    raise NotImplementedError
+  def postorder(current = @root, array = [])
+    if current == nil
+      return array
+    else
+      postorder(current.left, array)
+      postorder(current.right, array)
+      array << { key: current.key, value: current.value }
+    end
   end
 
   # Time Complexity: 
   # Space Complexity: 
-  def height
-    raise NotImplementedError
+  def height(current = @root, array)
```

Comment:
The height of a tree at any node will be the max height of the left subtree compared to the right subtree plus one.

Does that help?

Remember if a `current` is nil the height is 0.